File: client/src/app/encryption/computer.ts
import {Injectable} from 'angular2/core';
import {BigInteger} from 'jsbn';

import {EncryptionHelper} from './encryption_helper';

@Injectable()
export class Computer {

  private static operations = [
    '+', // add
    '-', // subtract
    '/', // divide
    '*', // multiply
    '%', // modulus
    '&', // powerMod
    '$'  // inverseMod
  ];

  private bitLength = 16;

  constructor(
    private encryptionHelper: EncryptionHelper
  ) {
  }

  public setBitLength(bits: number) {
    this.bitLength = bits;
  }

  public calculateStepCompute(compute: string, scope: {}): BigInteger {
    return this.doCommand(compute, scope);
  }

  private doCommand(command: string, scope: {}): BigInteger {

    let generateRandomPrimeRegex = /generateRandomPrime\(\)/;
    let generateRRegex = /generateR\((.+)\)/;
    let primitiveRootRegex = /primitiveRoot\((.+)\)/;
    let randomArbitraryRegex = /randomArbitrary\((.+),(.+)\)/;
    let findCoPrimeRegex = /findCoPrime\((.+)\)/;

    let matches;
    switch (true) {
      case generateRandomPrimeRegex.test(command):
        return this.encryptionHelper.generatePrime(this.bitLength);
      case generateRRegex.test(command):
        matches = generateRRegex.exec(command);
        return this.encryptionHelper.generateR(scope[matches[1]]);
      case primitiveRootRegex.test(command):
        matches = primitiveRootRegex.exec(command);
        return this.encryptionHelper.findPrimitiveRootOfPrime(scope[matches[1]]);
      case randomArbitraryRegex.test(command):
        matches = randomArbitraryRegex.exec(command);
        return this.encryptionHelper.getRandomArbitrary(
          this.resolveVariable(matches[1].trim(), scope).intValue(),
          this.resolveVariable(matches[2].trim(), scope).intValue()
        );
      case findCoPrimeRegex.test(command):
        matches = findCoPrimeRegex.exec(command);
        return this.encryptionHelper.findCoPrime(scope[matches[1]]);

      default:
        return this.calcExpression(command, scope);
    }

  }

  private findTopOperatorLocation(expr: string): number {
    let bracketCounter = 0;
    let topLevelLoc = -1;

    for (let i = 0; i < expr.length; i++) {
      let char = expr[i];
      if (char === '(') {
        bracketCounter++;
      } else if (char === ')') {
        bracketCounter--;
      } else if (bracketCounter === 0 && Computer.operations.indexOf(char) > -1) {
        topLevelLoc = i;
      }
    }

    if (bracketCounter !== 0 || topLevelLoc === -1) {
      throw new Error('Broken expression');
    }

    return topLevelLoc;
  }

  private trimBrackets(expr: string): string {
    if (expr[0] === '(' && expr[expr.length - 1] === ')') {
      return expr.substring(1, expr.length - 1);
    } else {
      return expr;
    }
  }

  private calcExpression(expr: string, scope: {}): BigInteger {
    // first split in to two, a and b by the top operator
    let topOpLoc = this.findTopOperatorLocation(expr);
    let topOp = expr[topOpLoc];

    let aStr = this.trimBrackets(expr.substring(0, topOpLoc).trim());
    let bStr = this.trimBrackets(expr.substring(topOpLoc + 1, expr.length).trim());

    let a;
    let b;
    let c;

    if (topOp === '&') {
      let parts = bStr.split(',');
      bStr = parts[0];
      let cStr = parts[1];

      // check if c requires more operations
      if (Computer.operations.some(function (v) {
          return cStr.indexOf(v) >= 0;
        })) {
        c = this.calcExpression(cStr, scope);
      } else {
        c = this.resolveVariable(cStr, scope);
      }
    }

    // check if a requires more operations
    if (Computer.operations.some(function (v) {
        return aStr.indexOf(v) >= 0;
      })) {
      a = this.calcExpression(aStr, scope);
    } else {
      a = this.resolveVariable(aStr, scope);
    }
    // check if b requires more operations
    if (Computer.operations.some(function (v) {
        return bStr.indexOf(v) >= 0;
      })) {
      b = this.calcExpression(bStr, scope);
    } else {
      b = this.resolveVariable(bStr, scope);
    }

    if (topOp === '&') {
      return this.calc(a, topOp, b, c);
    }
    return this.calc(a, topOp, b);
  }

  private calc(a: BigInteger, operator: string, b: BigInteger, c?: BigInteger) {

    switch (operator) {
      case '+':
        return a.add(b);
      case '-':
        return a.subtract(b);
      case '*':
        return a.multiply(b);
      case '/':
        return a.divide(b);
      case '%':
        return a.mod(b);
      case '$':
        return a.modInverse(b);
      case '&':
        return a.modPow(b, c);
      case '^':
        return a.pow(b.intValue());
      default:
        throw new RangeError('Cannot resolve operation: ' + operator);
    }
  }

  private resolveVariable(v: string, scope: {}): BigInteger {
    if (v in scope) {
      return scope[v];
    } else if (parseInt(v, 10) !== NaN) {
      return new BigInteger("" + v);
    } else {
      throw new RangeError('Cannot resolve variable: ' + v);
    }
  }

}

File: backend/HomomorphicEncryptionBackend/Encryption/Computer.py
"""
HomomorphicEncryptionBackend.Encryption.Computer
"""

from injector import inject
from HomomorphicEncryptionBackend.Manager.LogManager import LogManager

__author__ = "VJ Patel (vj@vjpatel.me)"


class Computer:
    """
    Computer
    """

    OPERATIONS = [
        "+",
        "-",
        "/",
        "*",
        "%",
        "&",
        "$"
    ]

    @inject(log_manager=LogManager)
    def __init__(self, log_manager):
        """
        :param LogManager log_manager:
        :return:
        """
        self.__logger = log_manager.get_logger("Computer")

    def compute(self,
                compute,
                scope):
        self.__logger.debug(compute)
        self.__logger.debug(scope)
        return self.__do_command(compute, scope)

    def __do_command(self, command, scope):
        return self.__calc_expression(command, scope)

    def __trim_brackets(self, expr):
        if expr[0] == "(" and expr[-1] == ")":
            return expr[1:-1]
        else:
            return expr

    def __calc_expression(self, expr, scope):
        top_op_loc = self.__find_top_operator_location(expr)
        top_op = expr[top_op_loc]

        self.__logger.debug("Top operator: {0} at: {1}".format(top_op, top_op_loc))

        a_str = self.__trim_brackets(expr[0:top_op_loc].strip())
        b_str = self.__trim_brackets(expr[top_op_loc + 1:].strip())

        self.__logger.debug("\taStr: {0}".format(a_str))
        self.__logger.debug("\tbStr: {0}".format(b_str))

        c = None
        if top_op == "&":
            parts = b_str.split(",")
            b_str = parts[0]
            c_str = parts[1]
            self.__logger("\tbStr: {0}".format(b_str))
            self.__logger("\tcStr: {0}".format(c_str))

            # resolve c
            c = self.__resolve_variable(c_str, scope)

        # check if a requires more operations
        if any(operator in a_str for operator in Computer.OPERATIONS):
            a = self.__calc_expression(a_str, scope)
        else:
            a = self.__resolve_variable(a_str, scope)

        # check if c requires more operations
        if any(operator in b_str for operator in Computer.OPERATIONS):
            b = self.__calc_expression(b_str, scope)
        else:
            b = self.__resolve_variable(b_str, scope)

        self.__logger.debug("\t\ta: {0}".format(a))
        self.__logger.debug("\t\tb: {0}".format(b))

        if top_op == "&" and c is not None:
            self.__logger.debug("\t\tc: {0}".format(c))
            self.__logger.debug(scope)

            return self.__calc(a, top_op, b, c)

        return self.__calc(a, top_op, b)

    def __calc(self, a, operator, b, c=None):
        if operator == "+":
            return a + b
        if operator == "-":
            return a - b
        if operator == "*":
            return a * b
        if operator == "/":
            return a // b
        if operator == "%":
            return a % b
        if operator == "$":
            return self.__modinv(a, b)
        if operator == "&":
            return pow(a, b, c)

    def __egcd(self, a, b):
        if a == 0:
            return b, 0, 1
        else:
            g, y, x = self.__egcd(b % a, a)
            return g, x - (b // a) * y, y

    def __modinv(self, a, m):
        g, x, y = self.__egcd(a, m)
        if g != 1:
            raise Exception('modular inverse does not exist')
        else:
            return x % m

    def __find_top_operator_location(self, expr):
        bracket_counter = 0
        top_level_loc = -1

        for i in range(0, len(expr)):
            char = expr[i]

            if char == "(":
                bracket_counter += 1
            elif char == ")":
                bracket_counter -= 1
            elif bracket_counter == 0 and char in Computer.OPERATIONS:
                top_level_loc = i

        if bracket_counter != 0 or top_level_loc == -1:
            raise Exception("Broken expression")

        return top_level_loc

    def __resolve_variable(self, v, scope):
        if v in scope:
            return int(scope[v])
        elif self.__is_int(v):
            return int(v)
        else:
            raise Exception("Cannot resolve variable: {0}".format(v))

    def __is_int(self, string):
        try:
            int(string)
            return True
        except ValueError:
            return False
